host/docs/uhd_image_loader.1
.TH "uhd_image_loader" 1 "3.9.0" UHD "User Commands"
.SH NAME
uhd_image_loader - UHD Image Loader

.SH DESCRIPTION
Burn firmware and FPGA images onto connected Ettus Research devices.

.SH SYNOPSIS
.B  uhd_image_loader [OPTIONS]

.SH OPTIONS
.IP "List options:"
--help
.IP "Device and loader arguments:"
--args=""
.IP "Custom firmware filepath:"
--fw-path=""
.IP "Custom FPGA filepath:"
--fpga-path=""
.IP "Don't burn firmware:"
--no-fw
.IP "Don't burn FPGA:"
--no-fpga

.SH SPECIFYING A PARTICULAR DEVICE
.sp
Without any arguments given through the \fIargs=""\fR option, this utility will start a session
with the first Ettus Research device it sees. The arguments shown below will narrow down the
utility's search for a particular device:

.SS All devices
.sp
The \fIname=\fR and \fIserial=\fR arguments can be used to specify any device except an
uninitialized USB device or an OctoClock bootloader. The \fItype=\fR argument can be used
to specify any device, as described below:

"type=" argument

Argument       |     Device

type=usrp2     |     USRP N200, USRP N210

type=b200      |     USRP B200, USRP B210

type=e3x0      |     USRP E310

type=octoclock |     OctoClock

.sp
NOTE: The USRP1, USRP2, and USRP B100 are not supported.

.SS Network devices
.sp
By default, this utility will check all network interfaces for network-based devices, but a specific
IP address can be specified with the \fIaddr=\fR argument.

.SS NI-RIO devices
.sp
An X-Series devices connected via PCIe can be specified by its NI-RIO resource through the \fIresource=\fR
argument, and the RPC port through which to communicate with it can be specified with the \fIrpc-port=\fR
argument. Using these options is not recommended, as their default values are almost always used.

.SS OctoClock devices
An OctoClock's name and serial are only exposed when the firmware is loaded, so if the device only has a
bootloader, only the \fIaddr=\fR argument can be used to find it.

.SH DEVICE-SPECIFIC LOADER OPTIONS
.sp
Certain devices have specific options for customizing their image loading process, and these can be passed
in through the \fI--args=""\fR option. These arguments are specified below:

.SS USRP N200, USRP N210
.sp
The \fIoverwrite-safe\fR option will overwrite the device's safe-mode firmware and FPGA images. This is
NOT RECOMMENDED, as these images serve as the backup if the device's primary images are corrupted.

.sp
The \fIreset\fR option will automatically reset the device once the loading process finishes. When the
device resets, it will have its new images loaded.

.SS USRP X300, USRP X310 (Ethernet only)
.sp
The \fIconfigure\fR option will automatically reset the device once the loading process finishes. When
the device resets, it will have its new FPGA image loaded.

.sp
The \fIverify\fR option will tell the device to internally verify the integrity of the image as it loads.
This greatly increases the loading time.

.SH EXAMPLES

.SS Load only the default FPGA image onto a specific N2x0 device and reset
.sp
uhd_image_loader --args="type=usrp2,addr=192.168.10.2,reset" --no-fw
.ft

.SS Load a custom FPGA image onto a specific X3x0 device
.sp
uhd_image_loader --args="type=x300,addr=192.168.40.2" --fpga-path="/home/user/my_x300_fpga_image.bit"
.ft

.fi

.SH SEE ALSO
UHD documentation:
.B http://files.ettus.com/manual/
.LP
GR-UHD documentation:
.B http://gnuradio.org/doc/doxygen/page_uhd.html
.LP
Other UHD programs:
.sp
uhd_images_downloader(1) usrp2_card_burner(1)
.SH AUTHOR
This manual page was written by Nicholas Corgan
for the Debian project (but may be used by others).
.SH COPYRIGHT
Copyright (c) 2015 Ettus Research LLC
.LP
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.
.LP
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.


CHANGELOG
Change Log for Releases
==============================

## 003.012.000.000
* N3x0: Add White Rabbit support, add N300 support, standard BIST
        includes fan, fix issue with 1GigE, switch to 2 radio blocks
        with 2 channels each, upgrade TDC to version 2.0, fix issue in
        ARM deframer
* X300: Enable BasicRX to use A/B/AB/BA muxing setups, more consistent
        logging, fix enumeration issue with TwinRX
* USRP2/N2x0: Re-add ability to modulate in the DAC, improve ISE
              settings to better meet timing
* B205mini: Fix global reset, improve timing in b205_ref_pll
* UHD: Remove a lot of Boost usage, mostly replaced by C++11 features,
       more unit tests, fix Boost 1.67 compatibility, fix compiler
       warnings, add API to query clock rate range, fix get_usrp_?x_info
* MPM: Refactored N3xx code, moved C++ standard to 14, refactor
       Boost.Python bindings, use CMake variable MPM_DEVICE
* Logging: Allow disabling fastpath msgs at runtime
* Docs: Clarified meaning of DSP frequencies, improved manual
        section on synchronization, added some known issues to B100,
        USRP2, and USRP1, update test test procedure description
* Examples: Improved benchmark_rate (added failure thresholds, fixed
  incorrect calculation of samples on drops, fixed timeout values),
  minor fixes to txrx_loopback_to_file
* Utils: Handle U's in calibration tools, create-lvbitx.py is now Py3k
         compatible, fixed git-hash.sh
* RFNoC: DDCs/DUCs use DDSes instead of CORDIC, add DMA-based replay
         block in FPGA, add 64-bit support to axi_wrapper, add compat
         number to radio block,
* Debian: Fix rules file, fix Changelog format
* Fix license headers
* This release includes all bugfixes and features from previous
  releases, in particular, the 3.11.* release cycle
* Known issues: N310: In some rare scenarios at high rates, streaming
  can fail and even bring the FPGA into a bad state.

## 003.011.001.000
* N310: fix compiler warnings
* N310: Implement digital loopback
* N3xx: Add N3xx devtest
* X300: Properly coerce master clock rate (tick rate)
* X300: handle bad weak_ptr during pcie discovery
* X300: handle bad weak_ptr during pcie discovery
* X300: Fix check_radio_config() to fix errors when using a single dboard
        in slot A
* B200: docs: Suggest modifying recv_frame_size for more stability
* B200: Fix bandwidth warnings and ranges
* N2xx: Fix regression issue that limited tuning range
* UBX: Change antenna functions to coercers on antenna/value properties
* adf4002: Fix register programming for power down bit
* UHD: Fix config file path for some Windows builds
* UHD: Add operators == and != for uhd::dict
* UHD: Add device_addr_t constructor from map
* UHD: Fix range of gain group to skip gains with zero step
* UHD: Changes to support Boost 1.67
* UHD: Correctly set end of burst flag in RX metadata
* UHD: Reduce usage of boost::assign, boost::this_thread::sleep, and boost:bind
* UHD: Update multi_usrp::get_usrp_?x_info() for MPM devices
* UHD: Refactor static const values to fix linker errors in niusrprio
* mpm: cmake: Add git hash and version info to Python module
* mpm: Add reference counters to UIO
* mpm: Add offset to EEPROM reads
* mpm: Disable PPS out during initialization
* mpm: Update cmake to find the correct python3
* mpm: Bump maximum supported revision to 5 (Rev F)
* mpm: Fixed db slot typo in db-id
* mpm: Increased claim timeout, made a separate RPC connection for claim, and
       added asyn calls for long RPC executions
* mpm: Improve xport<->SFP mapping algorithm
* mpmd: Improved find routine to fail fast and verify correct device is
        reachable
* mpmd: Add missing virtual destructors
* rfnoc/x300: Make sure peek32() and peek64() are called with actual addresses
* rfnoc: ctrl_iface cleanup
* rfnoc radio: Improve warning for too many samples requested
* rfnoc radio: get_rx_stream resets sequence num
* examples: Increase settling time, increase buffer fill time, and fix subdevice
            selection in txrx_loopback_to_file
* examples: Improvements to benchmark_rate
* utils: downloader supports multiple RegExs
* utils: Added code to handle underruns during self calibration
* utils: Fix 30s tiemout in query_gpsdo_sensors
* logging: Improve style consistency and demote some messages
* logging: Fix UHD_LOG_FILE cmake variable
* Docs: Add Known Issues section to USRP1, B100, and USRP2/N2x0
* Docs: Hide dependencies directory from Doxygen
* Docs: Clarify subdev specs and magnesium driver usage for N300/N310
* cmake: Improve warning for missing requests
* cmake: update NSIS template
* cmake: Remove images downloader section (replaced by manifest)

## 003.011.000.001
* N300: Added driver support (includes some refactoring of N3xx
  codebase)
* MPM: Fix PyLint warnings, Fix error handling for TCA communication
  errors, Fix printout of AD9371 version
* uhd_images_downloader: Create unique archive names for images archives
  (now include git hashes in the filename)
* uhd_images_downloader: Fix SHA256 check
* utils: Add support for N3xx filesystem images to images downloader
* UHD: Minor logging fixes
* UHD: fix legacy compat to work with 2TX radio block
* X300: improve lvbitx bitstream md5 read time
* examples: Enhance benchmark_rate with more stats and timestamps for errors
* cmake: Correctly fail when an unavailable component is requested
* debian: Add UHDConfig.cmake to install list for libuhd-dev

## 003.011.000.000
* N310: Added driver
* UBX: Add support for CAL antenna for Rev E
* Added the module peripheral manager (MPM) with all subcomponents (N310
  drivers, mpm_shell, RPC server, BIST, etc.)
* UHD: Added rpclib as a internally tracked dependency
* UHD: Reduced the usage of Boost
* UHD: Updated uhd_images_downloader: Now uses more elaborate manifest to
  optimize downloads
* UHD: Introduced uhdlib internal include paths
* UHD: Add support for configuration files. USRP settings can now be set using

host/utils/uhd_image_loader.cpp
//
// Copyright 2015-2017 Ettus Research
// Copyright 2018 Ettus Research, a National Instruments Company
//
// SPDX-License-Identifier: GPL-3.0-or-later
//

#include <csignal>
#include <cstdlib>
#include <iostream>

#include <boost/assign.hpp>
#include <boost/filesystem.hpp>
#include <boost/format.hpp>

#include <uhd/config.hpp>
#include <uhd/image_loader.hpp>
#include <uhd/types/device_addr.hpp>
#include <uhd/utils/safe_main.hpp>
#include <boost/program_options.hpp>

namespace fs = boost::filesystem;
namespace po = boost::program_options;

static std::string device_type = "";
static int num_ctrl_c = 0;

/*
 * If the user presses Ctrl+C, warn them that they may corrupt their device.
 * If they press it again, provide instructions on restoring the device
 * (if applicable) and exit.
 */
void sigint_handler(int){
    num_ctrl_c++;
    if(num_ctrl_c == 1){
        std::cout << std::endl
                  << "Are you sure you want to abort? If you do, your device will likely" << std::endl
                  << "be in an unstable or unusable state." << std::endl
                  << "Press Ctrl+C again to abort." << std::endl << std::endl;
    }
    else{
        std::cout << std::endl << uhd::image_loader::get_recovery_instructions(device_type) << std::endl;
        exit(EXIT_FAILURE);
    }
}

int UHD_SAFE_MAIN(int argc, char *argv[]){

    po::options_description desc("Allowed options");
    desc.add_options()
        ("help", "help message")
        ("args", po::value<std::string>()->default_value(""), "Device args, optional loader args")
        ("fw-path", po::value<std::string>()->default_value(""), "Firmware path (uses default if none specified)")
        ("fpga-path", po::value<std::string>()->default_value(""), "FPGA path (uses default if none specified)")
        ("out-path", po::value<std::string>()->default_value(""), "Output path/filename of the downloaded FPGA .bit file")
        ("no-fw", "Don't burn firmware")
        ("no-fpga", "Don't Burn FPGA")
        ("download", "Download an image to a bit/bin file")
    ;

    po::variables_map vm;
    po::store(po::parse_command_line(argc, argv, desc), vm);
    po::notify(vm);

    // Help message
    if (vm.count("help")){
        std::cout << "UHD Image Loader" << std::endl
                  << std::endl
                  << "Load firmware and/or FPGA images onto an Ettus Research device." << std::endl
                  << std::endl
                  << desc << std::endl;
        return EXIT_FAILURE;
    }

    // Convert user options
    uhd::image_loader::image_loader_args_t image_loader_args;
    image_loader_args.args          = vm["args"].as<std::string>();
    image_loader_args.load_firmware = (vm.count("no-fw")    == 0);
    image_loader_args.load_fpga     = (vm.count("no-fpga")  == 0);
    image_loader_args.download      = (vm.count("download") != 0);
    image_loader_args.firmware_path = vm["fw-path"].as<std::string>();
    image_loader_args.fpga_path     = vm["fpga-path"].as<std::string>();
    image_loader_args.out_path      = vm["out-path"].as<std::string>();

    // Force user to specify a device
    if(not image_loader_args.args.has_key("type")){
        throw uhd::runtime_error("You must specify a device type.");
    }

    // Clean up paths, if given
    if(image_loader_args.firmware_path != ""){
        #ifndef UHD_PLATFORM_WIN32
        if(image_loader_args.firmware_path.find("~") == 0){
            image_loader_args.firmware_path.replace(0,1,getenv("HOME"));
        }
        #endif /* UHD_PLATFORM_WIN32 */
        image_loader_args.firmware_path = fs::absolute(image_loader_args.firmware_path).string();
    }
    if(image_loader_args.fpga_path != ""){
        #ifndef UHD_PLATFORM_WIN32
        if(image_loader_args.fpga_path.find("~") == 0){
            image_loader_args.fpga_path.replace(0,1,getenv("HOME"));
        }
        #endif /* UHD_PLATFORM_WIN32 */
        image_loader_args.fpga_path = fs::absolute(image_loader_args.fpga_path).string();
    }
    if(image_loader_args.out_path != ""){
        #ifndef UHD_PLATFORM_WIN32
        if(image_loader_args.out_path.find("~") == 0){
            image_loader_args.out_path.replace(0,1,getenv("HOME"));
        }
        #endif /* UHD_PLATFORM_WIN32 */
        image_loader_args.out_path = fs::absolute(image_loader_args.out_path).string();
    }

    // Detect which type of device we're working with
    device_type = image_loader_args.args.get("type","");

    std::signal(SIGINT, &sigint_handler);
    if(not uhd::image_loader::load(image_loader_args)){
        std::cerr << "No applicable UHD devices found" << std::endl;
        return EXIT_FAILURE;
    }

    return EXIT_SUCCESS;
}
